src/jit/morph.cpp

```diff
@@ -6702,11 +6702,22 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
                 }
                 else
                 {
+                    // In R2R mode the field offset for some fields may change when the code
+                    // is loaded. So we can't rely on a zero offset here to suppress the null check.
+                    //
+                    // See GitHub issue #16454.
+                    bool fieldHasChangeableOffset = false;
+
+#ifdef FEATURE_READYTORUN_COMPILER
+                    fieldHasChangeableOffset = (tree->gtField.gtFieldLookup.addr != nullptr);
```

Comment:
If we can’t trust the value of fldOffset for the R2R case it would seem better to handle it above with fgIsBigOffset as it could potentially be a “Big” offset as well.

Reply:
That case is likely covered by R2R pitching the prejitted code when a changeable offset ends up being a big offset (see the `ENCODE_FIELD_OFFSET` case in `LoadDynamicInfoEntry`).

Though thinking about it some more I wonder if we could get burned by cases where we have nested structs such that any particular offset is not big but the cumulative offset after R2R fixups have been applied is big. Let me try working up a test case for this.

We should review who else calls `fgIsBigOffset` and see if those callers can be fooled by fields with changeable offsets.